### ISSUE TITLE
feat: support for rebase tokens (including interest bearing tokens)

### ISSUE THREAD

**Issue by xenide** [vexchange/v3-core#118]
## Scope 

We want to support rebase tokens for both `ConstantProductPair` and `StablePair`. If we're going to support this for `StablePair`, the marginal cost for supporting `ConstantProductPair` is not high 
 
To achieve this, the following changes are needed: 

- remove reserve0 / reserve1 as the basis for figuring out how many tokens are in the pair at any one time
- all contracts interacting with the pair have to implement a callback to transfer the tokens after reading in the balance at that point in time 
   - this adds 2 external calls to the workflow: `balanceOf` for both tokens 
- for mint: 
   - `reservoirMintCallback` which transfers tokens into the pair after the user receives the LP tokens 
   - our router would have to implement this 
- for swap: 
   - `reservoirSwapCallback`, which transfers tokens into the pair after user receives the output tokens 
   - our router would have to implement this 
- for burn: 
   - `reservoirBurnCallback`
   - do external call to read token balances 
   - actually for burn there is no need for a callback? The LP token can be transferred into the pair beforehand no problem, since the LP token isn't going to rebase

---

Curve / saddle supports this by: 
- pulling the swap amount after reading the balances in the pool/pair 
- for them they don't have a router so the pair/pool doing the pulling makes sense. We have a router so we have to use callbacks. Also I don't think killing the router makes sense for us 

Balancer: 
- doesn't support native rebase tokens, only wrappers of it 

Trader Joe: 
- doesn't support rebase for both v1 and the new liquidity book (v2) 
- the AMPL-AVAX / TIME-AVAX pool is just a normal pair 

Sushiswap: 
- doesn't seem to support it natively 
- the rebase library they have only supports bento shares <--> underlying token calculation

**Comment by OliverNChalk** [comment]
An alternative design:

Rebase delegations:

A wrapper token that instead of accruing the rebase into the share price enables a user (pair) to issue a unique wrapper token that is a claim on 1 unit of the underlying. Any positive/negative interest is then the responsbility of the issuer (pair).

This means in the interest case the pair just needs to be able to receive the interest through some other mechanism. In the negative interest case then we have a potential bank-run to manage through frequent reporting (and forced reporting on any `mint`/`/burn`.

Pros:
- Doesn't change the fundamental swap flow

Cons:
- Could be quite a bit more expensive for users that wrap/unwrap on every swap
- Moves complexity out of the router/pair and into a new, potentially unaudited, contract
- If everyone wraps/unwraps on every swap, it might be better to take the UX hit and use callbacks

**Comment by xenide** [comment]
After some discussion, we came up with an alternative to the callback. The pair can always pull from `msg.sender` which in most cases is the router. Then the router would have to pull the funds from the user before the start of the first swap, and would have to hold all intermediate tokens before being pulled for the next swap. The router would also have to approve each pair for unlimited spend of the tokens. We thought that this would be more gas efficient than the callback as it would avoid an external call to the router. 

Then we had a discussion whether we want to support this feature at this stage. We had 3 options
1. Force all pairs to use callback/more complex transfer flow
2. Force only StablePair to use it
3. Force no pairs to use it, introduce a new curve later if we see strong demand

We decided on option 3 as there might not be a lot of use case for it right now, and investing a lot of time on it might not be the best use of our time. There would be some complexity in the router if we choose to support it in the future, but this complexity should be manageable.
